File: src_local/agents/Support/Analysis/coverage_auditor.ts
import winston from "winston";
import { Path } from "../../../core/path_utils.ts";

const logger = winston.child({ module: "CoverageAuditor" });

/**
 * 📐 Auditor de Cobertura PhD (Bun Version).
 */
export class CoverageAuditor {
    detectTest(filePath: Path, compType: string, allFiles: string[], fInfo: any = null): boolean {
        if (this.isExemptFromTesting(compType, fInfo)) return true;

        const cleanStem = this.getCleanStem(filePath.name());
        const normalizedStem = cleanStem.replace(/_/g, "");

        for (const fileName of allFiles) {
            if (this.isTestMatch(fileName, normalizedStem, compType, filePath.name())) {
                return true;
            }
        }

        return false;
    }

    private isExemptFromTesting(compType: string, fInfo: any): boolean {
        const type = compType || "UNKNOWN";
        const isBoilerplate = fInfo ? (fInfo.complexity || 1) <= 1 : true;

        if (type === "DOC" || type === "TEST") return true;
        return ["CONFIG", "PACKAGE_MARKER", "UTIL"].includes(type) && isBoilerplate;
    }

    private getCleanStem(fileName: string): string {
        const rawStem = fileName.split('.')[0]!.toLowerCase();
        return rawStem.replace(/_persona$/, "").replace(/_agent$/, "").replace(/_engine$/, "");
    }

    private isTestMatch(fileName: string, normalizedStem: string, compType: string, originalName: string): boolean {
        const lowName = fileName.toLowerCase();
        const isTestFile = lowName.startsWith("test_") || lowName.endsWith(".test.ts") || lowName.endsWith(".spec.ts");

        if (!isTestFile) return false;

        const normalizedFile = lowName.replace(/_/g, "");
        if (normalizedFile.includes(normalizedStem)) {
            logger.debug(`Test match found: ${fileName} for ${originalName}`);
            return true;
        }

        return compType === "AGENT" && lowName.includes(normalizedStem);
    }
}

File: src_native/analyzer/src/coverage_auditor.rs
use serde::{Deserialize, Serialize};
use std::path::Path;

#[derive(Debug, Serialize, Deserialize)]
pub struct CoverageRequest {
    pub file_path: String,
    pub component_type: String,
    pub all_files: Vec<String>,
    pub complexity: Option<f64>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CoverageResponse {
    pub has_test: bool,
}

pub fn detect_test(request: CoverageRequest) -> CoverageResponse {
    if is_exempt_from_testing(&request.component_type, request.complexity.unwrap_or(1.0)) {
        return CoverageResponse { has_test: true };
    }

    let file_name = Path::new(&request.file_path)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or("");
    
    let clean_stem = get_clean_stem(file_name);
    let normalized_stem = clean_stem.replace('_', "");

    for other_file in &request.all_files {
        if is_test_match(other_file, &normalized_stem, &request.component_type, file_name) {
            return CoverageResponse { has_test: true };
        }
    }

    CoverageResponse { has_test: false }
}

fn is_exempt_from_testing(comp_type: &str, complexity: f64) -> bool {
    let is_boilerplate = complexity <= 1.0;
    if comp_type == "DOC" || comp_type == "TEST" { return true; }
    vec!["CONFIG", "PACKAGE_MARKER", "UTIL"].contains(&comp_type) && is_boilerplate
}

fn get_clean_stem(file_name: &str) -> String {
    let raw_stem = file_name.split('.').next().unwrap_or("").to_lowercase();
    raw_stem
        .replace("_persona", "")
        .replace("_agent", "")
        .replace("_engine", "")
}

fn is_test_match(file_name: &str, normalized_stem: &str, comp_type: &str, original_name: &str) -> bool {
    let low_name = file_name.to_lowercase();
    let is_test_file = low_name.starts_with("test_") || low_name.ends_with(".test.ts") || low_name.ends_with(".spec.ts");

    if !is_test_file { return false; }

    let normalized_file = low_name.replace('_', "");
    if normalized_file.contains(normalized_stem) {
        return true;
    }

    // Agent specialized matching
    comp_type == "AGENT" && low_name.contains(normalized_stem)
}
